design/clean_artifacts_in_csi_flow.md

```diff
@@ -0,0 +1,487 @@
+# Design to clean the artifacts generated in the CSI backup and restore workflows
+
+## Terminology
+
+* VSC: VolumeSnapshotContent
+* VS: VolumeSnapshot
+
+## Abstract
+* The design aims to delete the unnecessary VSs and VSCs generated during CSI backup and restore process. 
+* The design also make the VSC and VS restore action using Async Operations to wait for them turing `ReadyToUse` before returning.
+    * The advantage is the Restore can detect the VS and VSC creation error during restore.
+    * It can avoid restore completes successfully, but the volumes generated from the restored VS and VSC are not usable.
+
+## Background
+In the current CSI backup and restore workflows, please notice the CSI B/R workflows means only using the CSI snapshots in the B/R, not including the CSI snapshot data movement workflows, some artifacts are kept after the backup or the restore process.
+
+Some of them are kept due to design, for example, the VolumeSnapshotContents generated during the backup are kept to make sure the backup deletion can clean the snapshots in the storage providers.
+
+Some of them are kept by accident, for example, after restore, two VolumeSnapshotContents are generated for the same VolumeSnapshot. One is from the backup content, and one is dynamically generated from the restore's VolumeSnapshot.
+
+The design aims to clean the unnecessary artifacts, and make the CSI B/R workflow more concise and reliable.
+
+## Goals
+- Clean the redundant VSC generated during CSI backup and restore.
+- Remove the VSCs in the backup sync process.
+- Change the VS restore process to async.
+
+## Non Goals
+- There were some discussion about whether Velero backup should include VSs and VSCs not generated in during the backup. By far, the conclusion is not including them is a better option. Although that is a useful enhancement, that is not included this design.
+- Delete all the CSI-related metadata files in the BSL is not the aim of this design. 
+
+## Detailed Design
+### Backup
+During backup, the main change is the backup-generated VSCs should not kept anymore.
```

Comment:
If VSC resources are deleted, wouldn't that result in underlying CSI snapshots to be deleted? I always thought that the deletion policy only applies to the relationship between VS and VSC and if VSC is deleted, corresponding CSI snapshot is always deleted. May be I am missing something here?

Reply:
Please check the related external-snapshotter code.
https://github.com/kubernetes-csi/external-snapshotter/blob/7b580e5cb65e1b80e417b0094515d35a41799ba2/pkg/sidecar-controller/snapshot_controller.go#L56-L74

The Velero code always relies on deleting VSC, whose `DeletionPolicy` is set to `Retain`, not delete the snapshot.
The Velero will delete the VS and set the corresponding VSC's referenced VS to a non-existent one. That's in the VSC's spec, but VSC's spec cannot be modified. As a result, Velero needs to delete it first, then recreate it.
https://github.com/vmware-tanzu/velero/blob/0d18f1d156fad8a0555a6a236e606a47d7c4a566/pkg/util/csi/volume_snapshot.go#L580-L653

Reply:
VSC can be deleted without deleting cloud provider snapshot as long as driver supports it.

I believe that's how some were able to migrate using velero aws native snapshot to another cluster without datamover.

You can create a VSC from aws's snaphandle in a separate cluster and it would point to the same resource.

Reply:
Thanks @blackpiglet. It does seem that you can safely delete VSC but still keep the underlying snapshot, as long as "deletionPolicy" is "Retain". @kaovilai, I think the feature works for all drivers as it is implemented by the external snapshotter.

Reply:
Just some drivers have their quirks like azure file [would not keep a snapshot if its associated PVC is deleted](https://github.com/vmware-tanzu/velero/pull/8676#discussion_r1952110341). But generally considering only VSC yes.